src/langgraph_google_alloydb_pg/async_checkpoint.py

```diff
@@ -0,0 +1,352 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import asyncio
+import asyncpg # type: ignore
+
+from contextlib import asynccontextmanager
+
+import json
+from typing import List, Sequence, Any, AsyncIterator, Iterator, Optional, Dict, Tuple
+
+from sqlalchemy import text
+from sqlalchemy.ext.asyncio import AsyncEngine
+
+from langchain_core.runnables import RunnableConfig
+
+from langgraph.checkpoint.base import (
+    WRITES_IDX_MAP,
+    BaseCheckpointSaver,
+    ChannelVersions,
+    Checkpoint,
+    CheckpointMetadata,
+    CheckpointTuple,
+    get_checkpoint_id
+)
+from langgraph.checkpoint.serde.jsonplus import JsonPlusSerializer
+from langgraph.checkpoint.serde.types import TASKS, ChannelProtocol
+
+from langgraph.checkpoint.serde.base import SerializerProtocol
+
+MetadataInput = Optional[dict[str, Any]]
+
+from .engine import AlloyDBEngine
+
+
+class AsyncAlloyDBSaver(BaseCheckpointSaver[str]):
+    """Checkpoint stored in an AlloyDB for PostgreSQL database."""
+    
+    __create_key = object()
+    
+    def __init__(
+        self,
+        key: object,
+        pool: AsyncEngine,
+        schema_name: str = "public",
+        serde: Optional[SerializerProtocol] = None
+    ) -> None:
+        super().__init__(serde=serde)
+        if key != AsyncAlloyDBSaver.__create_key:
+            raise Exception(
+                "only create class through 'create' or 'create_sync' methods"
+            )
+        self.pool = pool
+        self.schema_name = schema_name
+        
+    @classmethod
+    async def create(
+        cls,
+        engine: AlloyDBEngine,
+        schema_name: str = "public",
+        serde: Optional[SerializerProtocol] = None
+    ) -> "AsyncAlloyDBSaver":
+        """Create a new AsyncAlloyDBSaver instance.
+
+        Args:
+            engine (AlloyDBEngine): AlloyDB engine to use.
+            schema_name (str): The schema name where the table is located (default: "public").
+            serde (SerializerProtocol): Serializer for encoding/decoding checkpoints (default: None).
+
+        Raises:
+            IndexError: If the table provided does not contain required schema.
+
+        Returns:
+            AsyncAlloyDBSaver: A newly created instance of AsyncAlloyDBSaver.
+        """
+        
+        checkpoints_table_schema = await engine._aload_table_schema("checkpoints", schema_name)
```

Comment:
Let's use a variable for the table name "checkpoints" and "checkpoint_writes" in the engine class and import them here, so we make sure they stay consistent